### PR TITLE
Peter/loans feat fetch prices on chain

### PR DESCRIPTION
**Pull request by peterslany**
Moving price fetching from UI to lib, replacing CoinGecko and using on-chain oracle as price source for LTV, collateralization and average thresholds computation.

Added `getCollateralInfo` method takes parameters `lendPositions`, `borrowPositions`, and `loanAssets`. All of these are returned from lib methods `getLendPositionsOfAccount`, `getBorrowPositionsOfAccount` and `getLoanAssets`, respectively. `getLoanAssets` was extended with call to oracle pallet to get asset's exchange rate against bitcoin which is exposed in the `LoanAsset` object. Therefore `getCollateralInfo` method can be run synchronously. This method computes and returns `LoanCollateralInfo`.
`LoanCollateralInfo` is interface consisting of following properties and methods that can be later used on UI side:
- `totalLentBtc` - total aggregated amount of all supplied assets in BTC value
- `totalCollateralBtc` - total aggregated amount of all supplied assets used as collateral in BTC value
- `totalBorrowedBtc` - total aggregated amount of all borrowed assets in BTC value
- `borrowLimitBtc` - total amount that can be borrowed until collateral threshold is met, denominated in BTC
- `ltv` - loan to value 
- `collateralThresholdWeightedAverage` - Average collateral threshold. Denominated in decimals. 
- `liquidationThresholdWeightedAverage` - Average liquidation threshold. Denominated in decimals.
- `calculateBorrowLimitBtcChange` - helper function to compute values based on user's input - to replace similar helper on UI
- `calculateLtvAndThresholdsChange` -helper function that computes new LTV and thresholds based on user's action and input. Replacement for helper on UI.

Note:
All of these properties and methods are equivalent to what is being computed on UI side and should be used to replace UI logic. 

**Next steps on the UI side:**
- Replace logic on UI side with call to `getCollateralInfo` which already pre-computes all the needed values.
- Convert BTC values to USD using price from CoinGecko - should be only for preview purposes

### CODE REVIEW

File: src/parachain/loans.ts [hunk 1]
```diff
@@ -435,18 +540,29 @@ export class DefaultLoansAPI implements LoansAPI {
         return newMonetaryAmount(amount, rewardCurrency);
     }
 
+    async _getExchangeRate(fromCurrency: CurrencyExt): Promise<ExchangeRate<WrappedCurrency, CurrencyExt>> {
+        const wrappedCurrencyId = this.api.consts.escrowRewards.getWrappedCurrencyId;
+        const wrappedCurrency = tokenSymbolToCurrency(wrappedCurrencyId.asToken);
+        if (isCurrencyEqual(fromCurrency, wrappedCurrency)) {
+            const wrappedCurrencyToBitcoinRate = Big(1); // Use exchange rate of 1 between wrapped BTC and real BTC.
```
Comment (daniel-savu) [inline src/parachain/loans.ts:547]: > // Use exchange rate of 1 between wrapped BTC and real BTC.

As far as I can see, `isCurrencyEqual` checks for ticker equality. In which case would `fromCurrency` or `wrappedCurrency` be actual BTC?

File: src/types/loans.ts [hunk 2]
```diff
@@ -15,6 +15,22 @@ interface BorrowPosition extends LoanPosition {
     accumulatedDebt: MonetaryAmount<CurrencyExt>;
 }
 
+type LoanAction = "lend" | "withdraw" | "borrow" | "repay";
```
Comment (daniel-savu) [inline src/types/loans.ts:18]: What's the overall approach to using this `LoanAction` type? I can see that the amounts in `LoanCollateralInfo` are updated based on it, is that meant to just show quick visual feedback after a tx is sent? For instance, if a "liquidation" action occurs (or if the user withdraws/borrows etc with a script), those changes wouldn't be caught here

File: src/types/loans.ts [hunk 3]
```diff
@@ -15,6 +15,22 @@ interface BorrowPosition extends LoanPosition {
...
+interface LoanCollateralInfo {
```
Comment (daniel-savu) [inline src/types/loans.ts:19]: This interface contains information about a lot of aspects of the lending protocol, and `LoanCollateralInfo` is misleasing. I'd rename more generically to just `LoansStatus` or similar.

File: src/parachain/loans.ts [hunk 4]
```diff
@@ -335,6 +369,77 @@ export class DefaultLoansAPI implements LoansAPI {
         return this._getPositionsOfAccount(accountId, this._getBorrowPosition.bind(this));
     }
 
+    getLoanCollateralInfo(
```
Comment (daniel-savu) [inline src/parachain/loans.ts:372]: Also rename here to match the new name of the return type

File: src/parachain/loans.ts [hunk 5]
```diff
@@ -435,18 +540,29 @@ export class DefaultLoansAPI implements LoansAPI {
...
         const underlyingCurrency = await currencyIdToMonetaryCurrency(this.api, underlyingCurrencyId);
 
-        const [lendApy, borrowApy, [totalLiquidity, availableCapacity, totalBorrows], rewardCurrency] =
+        const [lendApy, borrowApy, [totalLiquidity, availableCapacity, totalBorrows], rewardCurrency, exchangeRate] =
             await Promise.all([
                 this._getLendApy(underlyingCurrencyId),
                 this._getBorrowApy(underlyingCurrencyId),
                 this._getTotalLiquidityCapacityAndBorrows(underlyingCurrency, underlyingCurrencyId),
                 this._getRewardCurrency(),
+                this._getExchangeRate(underlyingCurrency),
```
Comment (daniel-savu) [inline src/parachain/loans.ts:565]: Will this become the only place where oracle prices are fetched in the UI?

File: test/unit/parachain/loans.test.ts [hunk 6]
```diff
@@ -128,4 +134,173 @@ describe("DefaultLoansAPI", () => {
...
+            ).to.be.true;
+        });
+
+        it("should compute correct LTV and average thresholds", () => {
+            const governanceCurrencyLentAmount = Big(100);
+            const wrappedCurrencyLentAmount = Big(2);
+            const relayCurrencyLentAmount = Big(1000); // Should not affect the computation since it won't be enabled as collateral.
+
+            const governanceCurrencyBorrowedAmount = Big(10);
+            const relayCurrencyBorrowedAmount = Big(10);
```
Comment (daniel-savu) [inline test/unit/parachain/loans.test.ts:249]: Here too, please use `MonetaryAmount` instead of `Big`. The signatures of `mockLendPosition` and `mockBorrowPosition` should be easy to change so you don't have to unroll the object when calling the functions

File: src/parachain/loans.ts [hunk 7]
```diff
@@ -335,6 +369,77 @@ export class DefaultLoansAPI implements LoansAPI {
...
     }
 
+    getLoanCollateralInfo(
+        lendPositions: Array<LendPosition>,
+        borrowPositions: Array<BorrowPosition>,
+        loanAssets: TickerToData<LoanAsset>
+    ): LoanCollateralInfo {
+        const lendCollateralPositions = lendPositions.filter(({ isCollateral }) => isCollateral);
+        const lendCollateralThresholdAdjustedPositions = lendCollateralPositions.map((position) => {
+            const collateralTheshold = loanAssets[position.currency.ticker].collateralThreshold;
```
Comment (daniel-savu) [inline src/parachain/loans.ts:379]: > loanAssets[position.currency.ticker]

There's a chance this will throw an error if the currency ticker changes (more likely for foreign assets). It's unlikely, but you should add a doc comment about it at least, so the UI wraps this in a try/catch.

Comment (daniel-savu) [reply src/parachain/loans.ts:379]: This can also occur if the `loanAssets` parameter is empty (for whatever reason) while `lendCollateralPositions` is non-empty

File: test/unit/parachain/loans.test.ts [hunk 8]
```diff
@@ -128,4 +134,173 @@ describe("DefaultLoansAPI", () => {
...
+        });
+
+        it("should not throw when there are no positions", () => {
+            expect(loansApi.getLoanCollateralInfo([], [], loanAssets)).to.not.throw;
+        });
+
+        it("should not throw when there are no borrow positions", () => {
+            const lendPositions = [mockLendPosition(testGovernanceCurrency, 1)];
+            expect(loansApi.getLoanCollateralInfo(lendPositions, [], loanAssets)).to.not.throw;
+        });
```
Comment (daniel-savu) [inline test/unit/parachain/loans.test.ts:304]: You can add another test that passes empty a non-empty `lendPositions`, a non-empty `borrowPositions`, and an empty `TickerToData` for `loanAssets` to make sure it doesn't throw

File: src/parachain/loans.ts (hunk 1 above)
Comment (peterslany) [reply src/parachain/loans.ts:547]: Good point, the comment was misleading, will remove it. I meant that exchange rate is expressed against BTC and in case of wrapped BTC, the protocol assumes it has the exactly same value as BTC, right?

File: src/types/loans.ts (hunk 2 above)
Comment (peterslany) [reply src/types/loans.ts:18]: This type union is useful when UI needs to compute how the LTV metrics will change once `LoanAction` is done - before submitting the tx. It is used as in  `calculateBorrowLimitBtcChange`, `calculateLtvAndThresholdsChange`.

File: src/types/loans.ts (hunk 3 above)
Comment (peterslany) [reply src/types/loans.ts:19]: Naming this one is not easy, `LoansStatus` is already used on UI to denote risk of getting liquidated, so I would use `LoansAccountInfo`, what do you think about that one? Open to other suggestions!

File: src/parachain/loans.ts (hunk 5 above)
Comment (peterslany) [reply src/parachain/loans.ts:565]: No, the oracle price is used fetched in other places in lib that are used in UI too - vault capacity calculation, vault apy calculation.

File: src/parachain/loans.ts (hunk 1 above)
Comment (daniel-savu) [reply src/parachain/loans.ts:547]: >  the protocol assumes it has the exactly same value as BTC, right?

It does yeah

File: src/types/loans.ts (hunk 3 above)
Comment (daniel-savu) [reply src/types/loans.ts:19]: `LendingStats`?

File: src/parachain/loans.ts (hunk 5 above)
Comment (daniel-savu) [reply src/parachain/loans.ts:565]: For now it's fine but the UI could optimize this by fetching exchange rates just once and then passing them down to the various components, the same way `loanAssets` is passed in `getLoanCollateralInfo`. If you think this makes sense you can create a GitHub issue in the UI/lib.

File: src/types/loans.ts (hunk 3 above)
Comment (peterslany) [reply src/types/loans.ts:19]: sounds good :+1:

File: src/parachain/loans.ts (hunk 7 above)
Comment (peterslany) [reply src/parachain/loans.ts:379]: Actually, this function has to throw if `loanAssets` object does not contain data for some of the currencies that loan positions consist of. I added function to check this and return meaningful error message if such case occurs.

File: test/unit/parachain/loans.test.ts (hunk 8 above)
Comment (peterslany) [reply test/unit/parachain/loans.test.ts:304]: Added test to ensure that it _throws_, see my response to the previous comment.

File: src/parachain/loans.ts (hunk 5 above)
Comment (peterslany) [reply src/parachain/loans.ts:565]: I think it's not necessary at the moment since in the other place the exchange rate is fetched only for one currency - so fetching all extra exchange rates for loan currencies would not make the UI more efficient.

**Comment by peterslany** [comment]
@daniel-savu I edited the PR with your comments, please re-review the PR when you are back :pray:
